File: backend/designer/pricing.py
"""Deterministic, rule-based pricing for a design — no ML, no external calls.

Price = garment base price
      + sum of selected style-option deltas
      + a flat fee per print zone that has at least one layer in it
      + a flat fee per layer, by type (image artwork costs more to produce
        than plain text)
"""

from decimal import Decimal

ZONE_SETUP_FEE = Decimal('5.00')
LAYER_FEE_BY_TYPE = {
    'text': Decimal('2.00'),
    'image': Decimal('4.00'),
}


def compute_price(garment_type, selected_options: dict, layers: list) -> Decimal:
    price = Decimal(garment_type.base_price)

    if selected_options:
        option_lookup = {
            (opt.category, opt.key): opt.price_delta
            for opt in garment_type.style_options.all()
        }
        for category, key in selected_options.items():
            price += option_lookup.get((category, key), Decimal('0'))

    zones_used = {layer.get('zone') for layer in layers if layer.get('zone')}
    price += ZONE_SETUP_FEE * len(zones_used)

    for layer in layers:
        price += LAYER_FEE_BY_TYPE.get(layer.get('type'), Decimal('0'))

    return price.quantize(Decimal('0.01'))

File: frontend/src/lib/pricing.ts
import type { GarmentTypeDetail, DesignLayer } from '@/types/designer';

/**
 * Mirrors backend/designer/pricing.py exactly (same constants, same rule
 * order) so the studio shows an accurate live price before saving — the
 * server always recomputes and is the source of truth.
 */
const ZONE_SETUP_FEE = 150.0;
const LAYER_FEE_BY_TYPE: Record<string, number> = {
  text: 60.0,
  image: 120.0,
  // A pocket is cut, hemmed and stitched on rather than printed, so it's
  // priced as construction rather than as another screen.
  pocket: 180.0,
};

export function computePrice(
  garmentType: GarmentTypeDetail,
  selectedOptions: Record<string, string>,
  layers: DesignLayer[]
): number {
  let price = parseFloat(garmentType.base_price);

  for (const [category, key] of Object.entries(selectedOptions)) {
    const option = garmentType.style_options.find((o) => o.category === category && o.key === key);
    if (option) price += parseFloat(option.price_delta);
  }

  const zonesUsed = new Set(layers.map((l) => l.zone));
  price += ZONE_SETUP_FEE * zonesUsed.size;

  for (const layer of layers) {
    price += LAYER_FEE_BY_TYPE[layer.type] ?? 0;
  }

  return Math.round(price * 100) / 100;
}
